File: csharp/src/Training.Workshop/M6/S11SafeComposite/Step1/Combo.cs
using Training.Workshop.Shared;

namespace Training.Workshop.M6.S11SafeComposite.Step1;

/// <summary>Krok 1: zarządzanie dziećmi tylko w węźle; Describe() rozszerza opis o dzieci.</summary>
public sealed class Combo : MenuComponent
{
    private readonly string _name;
    private readonly List<MenuComponent> _children = [];

    public Combo(string name)
    {
        _name = name;
    }

    public override string Name => _name;

    public override Money Price
    {
        get
        {
            var total = Money.Zero;
            foreach (var child in _children)
            {
                total = total.Plus(child.Price);
            }
            return total;
        }
    }

    public void Add(MenuComponent child)
    {
        _children.Add(child);
    }

    public IReadOnlyList<MenuComponent> Children => _children.ToList().AsReadOnly();

    public override string Describe()
    {
        if (_children.Count == 0)
        {
            return base.Describe();
        }
        return base.Describe() + " [" + string.Join(", ", _children.Select(child => child.Describe())) + "]";
    }
}

File: typescript/src/workshop/m6/s11_safecomposite/step1/Combo.ts
import { Money } from '../../../shared/Money.js';
import { MenuComponent } from './MenuComponent.js';

/** Krok 1: zarządzanie dziećmi tylko w węźle; describe() rozszerza opis o dzieci. */
export class Combo extends MenuComponent {
  private readonly comboName: string;
  private readonly childList: MenuComponent[] = [];

  constructor(name: string) {
    super();
    this.comboName = name;
  }

  override name(): string {
    return this.comboName;
  }

  override price(): Money {
    let total = Money.ZERO;
    for (const child of this.childList) {
      total = total.plus(child.price());
    }
    return total;
  }

  add(child: MenuComponent): void {
    this.childList.push(child);
  }

  children(): readonly MenuComponent[] {
    return Object.freeze([...this.childList]);
  }

  override describe(): string {
    if (this.childList.length === 0) {
      return super.describe();
    }
    return `${super.describe()} [${this.childList.map((child) => child.describe()).join(', ')}]`;
  }
}
